### ISSUE TITLE
Missed violation for RequireThis check

### ISSUE THREAD

**Issue by bossever** [checkstyle/checkstyle#9164]
In reference to #7656, the below code (in example section of Javadoc) does not throw violation when checked by RequireThisCheck. All properties are used with their default values.
Documentation : https://checkstyle.org/config_coding.html#RequireThis

`$ cat test.java`
```
1  public class B {
2    private String field;
3  
4    public B(String f) {
5      field = f;
6    }
7  
8    public String addSuffixToField(String field) {
9    
10    // Overlapping by method argument.
11    return field += "suffix";   // Expected violation: Reference to instance variable "field" needs "this".
12    }
13   }
```

`$ cat custom_config.xml`
```
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="RequireThis">
      <property name="checkFields" value="true"/>  
      <property name="checkMethods" value="true"/>
      <property name="validateOnlyOverlapping" value="true"/>
    </module>
  </module>
</module>
      
```

`$ java -jar checkstyle-8.40-all.jar -c custom_config.xml Test.java`
```
Starting audit...
Audit done.
```

Expected violation on line 11: Reference to instance variable "field" needs "this". [RequireThis]
Is this as intended? Or a bug?

**Comment by nmancus1** [comment]
Note that the example from the documentation is a bit different than what you have posted:
```
public static class B {
    private int field; // type of class member is int

    public A(int f) {
      field = f;
    }

    String addSuffixToField(String field) {
        // Overlapping by method argument. Equal to "return field = field + "suffix";"
        return field += "suffix"; // violation: Reference to instance variable "field" needs "this".
    }
}
```
Adding `this` to `field` in the `addSuffixToField()` method will not even compile in this case. Please take care of this during your work on https://github.com/checkstyle/checkstyle/issues/7656.  I would make sure that all code examples compile (and the fixed violations of such code). 

It is great that you recognized this, however I believe this issue is a duplicate of https://github.com/checkstyle/checkstyle/issues/4216, however.

**Comment by nmancus1** [comment]
@pbludov @strkkk could #4216 AND #7656 be resolved in the scope of #7656?

**Comment by strkkk** [comment]
I think it is ok to fix both issues in one PR

**Comment by rnveach** [comment]
Unless the issues overlap similar code, I would try to keep issues separated, especially if one can be merged separately from the other. Big and extensive PRs slow the review down.

**Comment by nmancus1** [comment]
@rnveach both issues are for documentation, to improve the examples/ configs.

From #4216:
> All examples should be re-producible.
It should be made clear which config is being used by example.

From #7656:
> Add example for each existing config to show what code is ok and what code cause violation.

If anything, #4216 is just a detailed analysis of some of the work that should be done in #7656, do you agree?

**Comment by nmancus1** [comment]
Also, according to check logic, the code in the report shouldn't be a violation, in my opinion: `If method parameter is returned from the method, the check will not raise violation for
 returned variable/parameter...`

**Comment by bossever** [comment]
> Also, according to check logic, the code in this report shouldn't be a violation, in my opinion: `If method parameter is returned from the method, the check will not raise violation for
>  returned variable/parameter...`

Yes, I realised that a bit late, but why would the check implement such a logic if there is clearly an ambiguity as to which "field" field is being referred to while returning from the method?

**Comment by bossever** [comment]
> ```
> public static class B {
>     private int field; // type of class member is int
> 
>     public A(int f) {
>       field = f;
>     }
> 
>     String addSuffixToField(String field) {
>         // Overlapping by method argument. Equal to "return field = field + "suffix";"
>         return field += "suffix"; // violation: Reference to instance variable "field" needs "this".
>     }
> }
> ```
> Adding `this` to `field` in the `addSuffixToField()` method will not even compile in this case. Please take care of this during your work on https://github.com/checkstyle/checkstyle/issues/7656.  I would make sure that all code examples compile (and the fixed violations of such code). 


So this example must be removed right? Since the only violation stated in the code does not throw a violation, this code created a lot of confusion for me as https://github.com/checkstyle/checkstyle/issues/7656 was my first issue in this repo. Please guide me on how to go about it.

**Comment by nmancus1** [comment]
> this code created a lot of confusion for me

It did for me, as well :) Let's make it better. Regarding examples and configs, I would look at the [unit tests](https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheckTest.java) and corresponding test inputs.

For example...
For the default config, look at `testValidateOnlyOverlappingTrue()`:
https://github.com/checkstyle/checkstyle/blob/45ef1deb399b63a3bd968a8b4e411a9ac6d90517/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheckTest.java#L273-L290

We can see that no additional attributes are added (no `checkConfig.addAttribute()`, so this test uses the default config.  The string array `expected` shows all the expected violations for the test input. As you can see on line 289, the input file, `InputRequireThisValidateOnlyOverlappingTrue.java` is the file that we are checking in this test. [Take a look at this input file to see some examples that you can use](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/requirethis/InputRequireThisValidateOnlyOverlappingTrue.java).

**Comment by bossever** [comment]
@nmancus1 Okay, I will learn about unit testing and get back to this issue in a few hours. Thanks for the support.

**Comment by bossever** [comment]
> Also, according to check logic, the code in this report shouldn't be a violation, in my opinion: `If method parameter is returned from the method, the check will not raise violation for returned variable/parameter...`

@nmancus1 , I am closing this issue owing to this statement in the doc, thanks for the help and pointing to the tests, sorry for the delay!
